tiled/structures/awkward.py

```diff
@@ -0,0 +1,67 @@
+from dataclasses import dataclass
+
+import awkward
+
+
+@dataclass
+class AwkwardStructure:
+    length: int
+    form: dict
+
+    @classmethod
+    def from_json(cls, structure):
+        return cls(**structure)
+
+
+def project_form(form, form_keys_touched):
+    # See https://github.com/bluesky/tiled/issues/450
+    if isinstance(form, awkward.forms.RecordForm):
+        if form.fields is None:
+            original_fields = [None] * len(form.contents)
+        else:
+            original_fields = form.fields
+
+        fields = []
+        contents = []
+        for field, content in zip(original_fields, form.contents):
+            projected = project_form(content, form_keys_touched)
+            if projected is not None:
+                fields.append(field)
+                contents.append(content)
+
+        if form.fields is None:
+            fields = None
+
+        return form.copy(fields=fields, contents=contents)
+
+    elif isinstance(form, awkward.forms.UnionForm):
+        step1 = [project_form(x, form_keys_touched) for x in form.contents]
+        step2 = [x for x in step1 if x is not None]
+        if len(step2) == 0:
+            return None
+        elif len(step2) == 1:
+            return step2[0]
+        else:
+            raise NotImplementedError(
+                "Certain UnionForms are not yet supported. "
+                "See https://github.com/scikit-hep/awkward/issues/2666"
+            )
+            return awkward.forms.UnionForm.simplified(
+                form.tags,
+                form.index,
+                step2,
+                parameters=form.parameters,
+                form_key=form.form_key,
+            )
```

Comment:
@jpivarski @danielballan my understanding of the purpose of this routine is to build the form that reflects the array produced by a given slice. The benefit of doing this is to minimise reading.

This will be a problem for unions, because at present we want them to be erased if they're not actually required. This is once example of what we call "canonicalisation rules".

Consider a union of two records:
```
union[
   {x: int64, y: int32},
   {x: int64}
]
```
The following slice would erase the union
```python
array["x"]
```
In doing so, we'd copy and merge the buffers, i.e. we can't directly load the resulting array using `from_buffers`.

I think this presents a problem with the approach taken here.

In this instance, I have an idea; don't simplify the form *at all*, and use the private `simplify` argument to `ak.from_buffers`. This would change the form of the result, but it *would* do what we want here (despite concatenating buffers if unions with >1 content end up being erased).

We need to discuss this - @jpivarski I'll ping you in Slack.

Reply:
Applied in 109592c8c868210f76f5cb389551711fd7f6f922